### PR TITLE
feat(protocol): add repay bond amount calculation

### PR DESCRIPTION
**Pull request by scorpion9979**
Adds an on-chain getter method to the balance sheet to calculate the amount of bond needed to be repaid in order to liquidate a specified amount of a specified collateral.

### CODE REVIEW

**Comment by paulrberg** [comment]
I just started to review this and noticed that this PR does not test the functions it adds. I'm afraid you will have to test these, @scorpion9979 😬

A comment about `repayBondAmount`. That can be interpreted as as the number of bonds, because there is another storage variable in the protocol called `maxBonds`. I would rather define the parameter as simply `repayAmount`, since that variable is ubiquitously considered an amount of hTokens.

And a question about the logic itself. Didn't you copy-paste this from somewhere else? If yes, shouldn't we call this new function from that place, so that we keep the code base DRY?

File: packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol [hunk 1]
```diff
@@ -236,6 +236,41 @@ contract BalanceSheetV1 is
...
 
+    /// @inheritdoc IBalanceSheetV1
+    function getRepayBondAmount(
+        IHToken bond,
+        uint256 liquidationAmount,
+        IErc20 collateral
+    ) public view override returns (uint256 repayBondAmount) {
+        // Normalize the collateral amount.
+        uint256 normalizedLiquidationAmount;
+        uint256 collateralPrecisionScalar = 10**(18 - collateral.decimals());
```
Comment (paulrberg) [inline packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol:247]: This should be enclosed within an `unchecked` block to save gas. The number of decimals is [guaranteed](https://github.com/hifi-finance/hifi/blob/0eeff4c32b51d9f4ba6e26b369e04561297dbf4b/packages/protocol/contracts/core/fintroller/FintrollerV1.sol#L153-L155) to be 18 or lower.

**Review by paulrberg — CHANGES_REQUESTED** [review]

**Comment by scorpion9979** [comment]
> I just started to review this and noticed that this PR does not test the function it adds. I'm afraid you will have to write some tests, @scorpion9979 
> 
> A comment about the name of the `repayBondAmount` variable. It can be interpreted as as a number of bonds, because there is another storage variable in the protocol called `maxBonds`. I would rather define the parameter as simply `repayAmount`, since that variable is ubiquitously considered an amount of hTokens.
> 
> And a question about the logic itself. Didn't you copy-paste this from somewhere else? If yes, shouldn't we call this new function from that place, so that we keep the code base DRY?

I didn't actually do much copy-pasting. I basically used the math formula in `getSeizableCollateralAmount(...)`'s interface definition, but refactored it to solve for repay amount instead of collateral amount.

File: packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol (hunk 1 above)
Comment (scorpion9979) [reply packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol:247]: You mean `collateralPrecisionScalar`?
```solidity
uint256 collateralPrecisionScalar;
unchecked {
    collateralPrecisionScalar = 10**(18 - collateral.decimals());
}
```

If so, we need to do the same for `getSeizableCollateralAmount(...)`.

File: packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+          await this.mocks.wbtc.mock.decimals.returns(collateralDecimals);
+        });
+
+        it("retrieves the correct value", async function () {
+          const repayAmount: BigNumber = await this.contracts.balanceSheet.getRepayBondAmount(
+            this.mocks.hTokens[0].address,
+            liquidationAmount,
+            this.mocks.wbtc.address,
+          );
+          expect(repayAmount).to.equal(
```
Comment (scorpion9979) [inline packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: @paulrberg I added the unit tests for `getRepayBondAmount(...)`, but this particular line is failing when the number of decimals is 8. All other cases work just fine. I'm guessing this is due to bad math conversion either in the Solidity file or in the `mirrors.ts` function I added for the expected output, would be great if you could take a look.

File: packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol (hunk 1 above)
Comment (paulrberg) [reply packages/protocol/contracts/core/balanceSheet/BalanceSheetV1.sol:247]: Yes I mean `collateralPrecisionScalar`. And yes you're right, we need to do the same in `getSeizableCollateralAmount`. But let's not make that change in this PR.

File: packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts (hunk 2 above)
Comment (paulrberg) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: There were two issues:

1. You were using the `hUSDC` helper function to create a collateral amount.
2. You were not de-normalising the collateral amount (created mistakenly with `hUSDC`).

Comment (paulrberg) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: *Update*: I am still getting an error after correcting the errors above. There must be a third error.

Comment (paulrberg) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: Yup, the third error was in the Solidity code. You were using the PRBMath function `mul` on two values which were not normalised to 18 decimals. Only normalised amounts can be passed to PRBMath. You should have used the normal `*` operator.

Side note: this wouldn't have happened if we used the [typed version of UD60x18](https://github.com/hifi-finance/prb-math/blob/061ab63c61732b3e9ad037767be82874aec63094/contracts/PRBMathUD60x18Typed.sol). I am considering to switch over to that .. although this sort of bugs should always be caught by tests.

Comment (scorpion9979) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: Ok, so I've replaced `hUSDC` with `WBTC` helper function, replaced the `.mul` with the default Solidity multiplication, and added the unchecked math block.

Comment (paulrberg) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: Check Slack.

Comment (paulrberg) [reply packages/protocol/test/unit/balanceSheet/view/getRepayBondAmount.ts:82]: I fixed all of these on my local branch, plus more. I will have to git push with `--force`.

**Comment by paulrberg** [comment]
There is a thorough change log in my recent [commit message](https://github.com/hifi-finance/hifi/pull/38/commits/4d5648c67a2de74b32d5efef0f65d31070117ae5).

**Comment by paulrberg** [comment]
Besides the fixes we discussed above, I made two other big changes:

- [x] Remove the check for `liquidationIncentive == 1.0e18`. The value of the liquidation incentive is orthogonal to the calculation made in the `getRepayAmount` function, because the function *assumes* that there exists a seizable collateral amount that is greater than zero.
- [x] Rename `getRepayBondAmount` to `getBondAmount`, because `repayAmount` is implicitly an hToken amount.

**Pull request merged**